// @interacta/css-labels/src/styles.ts
import { css, injectGlobal } from '@emotion/css'
import { Padding, Options } from './types.js'

export const variables = injectGlobal`
  :root {
    --css-label-background-color: #1e2428;
    --css-label-brightness: brightness(150%);
  }
`

export const pointLabel = (padding: Padding, pointerEvents: Options['pointerEvents']): string => css`
  position: absolute;
  top: 0;
  left: 0;
  font-size: 14px;
  font-weight: 500;
  cursor: pointer;
  user-select: none;
  filter: var(--css-label-brightness);
  pointer-events: ${pointerEvents};
  background-color: var(--css-label-background-color);
  font-weight: 700;
  padding: ${padding.top}px ${padding.right}px ${padding.bottom}px ${padding.left}px;
  border-radius: 6px;
  
  transition: opacity 600ms;
  opacity: 0;
`

export const labelsContainer = css`
  transition: opacity 100ms;
  position: absolute;
  width: 100%;
  height: 100%;
  overflow: hidden;
  top: 0%;
  pointer-events: none;
  opacity: 1;
`

export const hidden = css`
  opacity: 0;

  div {
    pointer-events: none;
  }
}
`
